### PR TITLE
[rom_ext,rescue] Rescue protocol infrastructure

### PR DESCRIPTION
**Pull request by cfrantz**
1. Add function to get the mcycle timer to `ibex`.
2. Add UART receive functions to the `uart` driver.
3. Add `pinmux` support for UART receieve and reading strapping pins.

### CODE REVIEW

File: sw/device/silicon_creator/lib/drivers/pinmux.c [hunk 1]
```diff
@@ -114,6 +115,61 @@ static void enable_pull_down(top_earlgrey_muxed_pads_t pad) {
       kBase + PINMUX_MIO_PAD_ATTR_0_REG_OFFSET + pad * sizeof(uint32_t), reg);
 }
 
+/**
+ * Enables or disables pull-up/pull-down fro the specified pad.
```
Comment (pamaury) [inline sw/device/silicon_creator/lib/drivers/pinmux.c:119]: nit: "fro"

File: sw/device/silicon_creator/lib/drivers/pinmux.c [hunk 2]
```diff
@@ -114,6 +115,61 @@ static void enable_pull_down(top_earlgrey_muxed_pads_t pad) {
...
+ *
+ * @param pad A MIO pad.
+ * @param enable Whether the internal pull resistor should be enabled.
+ * @param up Whether the pull resistor should pull up(true) or down(false).
+ */
+static void enable_pull(top_earlgrey_muxed_pads_t pad, bool enable, bool up) {
```
Comment (pamaury) [inline sw/device/silicon_creator/lib/drivers/pinmux.c:125]: This function seems to subsume `enable_pull_down`. Should the old one be removed and its usage converted?

File: sw/device/silicon_creator/lib/drivers/pinmux.h [hunk 3]
```diff
@@ -23,6 +23,13 @@ extern "C" {
  */
 void pinmux_init(void);
 
+/**
+ * Read the SW_STRAP value considering weak/strong pull resistors.
+ *
+ * @return The strapping value 0-63.
+ */
+uint32_t pinmux_read_straps(void);
```
Comment (pamaury) [inline sw/device/silicon_creator/lib/drivers/pinmux.h:31]: Should this function document how the value is encoded (ie 2 bits per pin with weak/strong)?

File: sw/device/silicon_creator/lib/drivers/uart.h [hunk 4]
```diff
@@ -38,9 +51,32 @@ void uart_putchar(uint8_t byte);
...
 size_t uart_write(const uint8_t *data, size_t len);
 
+/**
+ * Read from the UART into a buffer.
+ *
+ * Reads up to `len` bytes into a buffer before the timeout.
+ *
+ * @param data Pointer to buffer to write.
+ * @param len Length of the buffer to write.
+ * @param timeout_ms The timeout to recieve the complete buffer.
```
Comment (pamaury) [inline sw/device/silicon_creator/lib/drivers/uart.h:63]: nit: "recieve"

**Review by pamaury — APPROVED** [review]
Overall looks good but I think the `ibex_mcycle` function already exists (see inlin comment).

File: sw/device/silicon_creator/lib/drivers/pinmux.c (hunk 1 above)
Comment (cfrantz) [reply sw/device/silicon_creator/lib/drivers/pinmux.c:119]: Fixed.

File: sw/device/silicon_creator/lib/drivers/uart.c [hunk 5]
```diff
@@ -86,6 +107,40 @@ size_t uart_write(const uint8_t *data, size_t len) {
...
+        return n;
+    }
+    uint32_t reg =
+        abs_mmio_read32(TOP_EARLGREY_UART0_BASE_ADDR + UART_RDATA_REG_OFFSET);
+    *data++ = (uint8_t)reg;
+  }
+  return n;
+}
+
+bool uart_break_detect(uint32_t timeout_us) {
```
Comment (moidx) [inline sw/device/silicon_creator/lib/drivers/uart.c:131]: Should we use a hardened boolean from the start given the potential for attack?

**Review by moidx — APPROVED** [review]

**Pull request merged**